node/workflow-api.md

```diff
@@ -0,0 +1,322 @@
+# 3rd revision of the Typescript SDK Workflow API
+
+### Current API
+
+```ts
+import { Trigger } from '@temporalio/workflow';
+
+export type MyWorkflow = (
+  arg1: number,
+  arg2: string
+) => {
+  execute(): Promise<void>;
+  signals: {
+    unblock(): void;
+  };
+  queries: {
+    isBlocked(): boolean;
+  };
+};
+
+export const myWorkflow: MyWorkflow = (arg1, arg2) => {
+  const unblocked = new Trigger<void>();
+  let isBlocked = true;
+  return {
+    // Main workflow logic
+    async execute() {
+      await unblocked;
+      isBlocked = false;
+    },
+    signals: {
+      unblock() {
+        unblocked.resolve();
+      },
+    },
+    queries: {
+      isBlocked() {
+        return isBlocked;
+      },
+    },
+  };
+};
+```
+
+#### Issues
+
+- Types must be explicitly declared (To make [eslint happy](https://github.com/typescript-eslint/typescript-eslint/blob/master/packages/eslint-plugin/docs/rules/explicit-module-boundary-types.md))
+- Very verbose
+- Trigger (ResolvablePromise) is a foreign concept to JS developers
+- Trigger doesn't provide all of the functionality for this workflow, an additional `isBlocked` boolean variable is required
+- Signals and queries must be defined at the top level in order for types to be inferred.
+
+#### Advantages
+
+- Types for all Workflow handlers can be inferred at Workflow handle creation
+
+### Goals for the redesign
+
+- Maintain type inference capabilities - Mandatory for the Typescript SDK
+- Dynamic signals and queries registration first - A single API for Workflow handler registration
+- Componentization of related functionality - increase reusability
+- Implicit interface declarations - less boilerplate
+- No global variables - to support running Workflows in less isolated runtimes
+
+### New proposal
+
+#### Remove the `Trigger` API in favor of `condition`
+
+`condition` is similar to `Workflow.await` in Java, it returns a Promise and accepts a closure which resolve the Promise when it returns `true`.
+
+```ts
+import { condition } from '@temporalio/workflow';
+
+let isBlocked = true;
+await condition(() => !isBlocked);
+```
+
+#### Remove the Workflow factory wrapper and always use dynamic handler registration
+
+`workflows.ts`
+
+```ts
+import { Signal, Query, condition } from '@temporalio/workflow';
+
+// Define the signal / query with the type signature at the top level to share the definition for handle creation
+export const unblockSignal = new Signal<[] /* args */>('unblock');
+export const isBlockedQuery = new Query<[] /* args */, boolean /* return type */>('isBlocked');
+
+// Types are inferred without an additional type definition
+export async function myWorkflow(arg1: number, arg2: string): Promise<void> {
+  let isBlocked = true;
+  unblockSignal.handler = () => (isBlocked = false);
+  isBlockedQuery.handler = () => isBlocked;
```

Comment:
Are these two handlers mutating a global field on `unblockSignal` and `isBlockedQuery`? Or is there some internal mechanism to have it specific to this workflow? What if `myWorkflow` was called concurrently, any clashes here? Does there need to be a `unblockedSignal.forWorkflow(this).handler = `? (pardon my ignorance on the internals of how global state is managed per workflow).

Reply:
It will be implemented with a setter that is workflow context aware

Reply:
Thanks! Curious how that setter might look. Mostly curious how one determines the workflow they are running in from unrelated code (no real thread-local concept in nodejs and I would think inspecting the stack may be brittle).

Reply:
Node.js has https://nodejs.org/api/async_context.html#async_context_class_asynclocalstorage, we implement a simple version of that for the workflow runtime too:
- https://github.com/temporalio/sdk-node/blob/7b80995de755d7ea098a235755892e871a1499b3/packages/workflow/src/async-local-storage.ts
- https://github.com/temporalio/sdk-node/blob/7b80995de755d7ea098a235755892e871a1499b3/packages/workflow/src/promise-hooks.ts
- https://github.com/temporalio/sdk-node/blob/7b80995de755d7ea098a235755892e871a1499b3/packages/worker/workflow-isolate-extension.cc

Reply:
I see, thanks! That is a good bit of magic for the developer. May also be worth exposing to the developer some static call that can inform them what workflow they are executing within if we don't already.

Reply:
i'd love for `isBlockedQuery` to just have a default handler, so i dont have to write this boilerplate, perhaps as part of queryable state.

Reply:
@sw-yx I'll add `QueryableState` in the next proposal.
I think we might even be able to make it into something like `StateVariable` where we auto generate signal handlers too.